client.go

```diff
@@ -40,35 +40,69 @@ func init() {
 
 // StatsdClient is a client library to send events to StatsD
 type StatsdClient struct {
-	conn             net.Conn
-	addr             string
-	prefix           string
-	eventStringTpl   string
-	Logger           Logger
-	conn_type        string
-	reconnect_ticker *time.Ticker
+	conn            net.Conn
+	addr            string
+	prefix          string
+	eventStringTpl  string
+	Logger          Logger
+	connType        string
+	reconnectTicker *time.Ticker
 }
 
-// NewStatsdClient - Factory
-func NewStatsdClient(addr string, prefix string) *StatsdClient {
+// ConfigurationFunc is a typedef for a function that configures some aspect of
+// a StatsdClient instance
+type ConfigurationFunc func(*StatsdClient)
+
+// AutoReconnect returns a ConfigurationFunc that causes the StatsdClient to automatically
+// recreate its underlying connection on the specified interval.
+func AutoReconnect(interval time.Duration) ConfigurationFunc {
+	return func(client *StatsdClient) {
+		if interval > 0 {
+			client.reconnectTicker = time.NewTicker(interval)
+		}
+	}
+}
+
+// UseLogger returns a ConfigurationFunc that makes the StatsdClient use the specified Logger
+// implementation, rather than the default implementation.
+func UseLogger(logger Logger) ConfigurationFunc {
+	return func(client *StatsdClient) {
+		client.Logger = logger
+	}
+}
+
+// NewStatsdClient is a factory func that creates a StatsdClient that sends to
+// the configured address and prefixes all stats with the given prefix name.
+func NewStatsdClient(addr string, prefix string, options ...ConfigurationFunc) *StatsdClient {
 	// allow %HOST% in the prefix string
 	prefix = strings.Replace(prefix, "%HOST%", Hostname, 1)
 	client := &StatsdClient{
-		addr:             addr,
-		prefix:           prefix,
-		Logger:           log.New(os.Stdout, "[StatsdClient] ", log.Ldate|log.Ltime),
-		eventStringTpl:   "%s%s:%s",
-		reconnect_ticker: time.NewTicker(30 * time.Second),
+		addr:           addr,
+		prefix:         prefix,
+		eventStringTpl: "%s%s:%s",
 	}
 
-	go func() {
-		for range client.reconnect_ticker.C {
-			err := client.Reconnect()
-			if err != nil {
-				fmt.Println(err)
+	// apply all the configuration functions
+	for _, opt := range options {
+		opt(client)
+	}
+
+	// set some defaults, if necessary
+	if client.Logger == nil {
+		client.Logger = log.New(os.Stdout, "[StatsdClient] ", log.Ldate|log.Ltime)
+	}
+
+	if client.reconnectTicker != nil {
+		go func() {
```

Comment:
@Jonnymcc   If I'm reading this correctly, it'll call this goroutine whenever reconnectTicker is set.  And with the `for range..`, it'll call reconnect() every ticks.  But.. I thought we want to call this in the event if the connection fails.  [Conn.write()](https://golang.org/pkg/net/#Conn) does return an error.  Can't we just call reconnect when write() returns error?

Reply:
Good question.

Reply:
1. I like the idea of reconnecting on a write error, but will there be an error if we are sending UDP? Also, our nodejs statsd client reconnects on a set interval I believe.

2. I'm not a go expert, but I would imagine that when a new conn object is created the old conn object would get garbage collected since there are no refs and would be "closed" appropriately.

Reply:
@Jonnymcc No, we need to close them ourselves. Go does not have destructors like C++ does.

Reply:
Good eye, @jsongHBO. We need to fix that. Could be a huge resource leak.

Reply:
So the previous conn needs to be closed before attempting to create the new one?

Reply:
@Jonnymcc   This is what the conn's close code looks like.. (sorry, I don't know how to provide link with line number)
```
// Close closes the connection.
func (c *conn) Close() error {
	if !c.ok() {
		return syscall.EINVAL
	}
	err := c.fd.Close()
	if err != nil {
		err = &OpError{Op: "close", Net: c.fd.net, Source: c.fd.laddr, Addr: c.fd.raddr, Err: err}
	}
	return err
}
```
so it seems that `fd` maybe "file descriptor"?  If so, then we definitely need to close it.

Reply:
yes.. netFD is file descriptor  https://golang.org/search?q=netFD

Reply:
> will there be an error if we are sending UDP

I don't know... I think underneath, the udp connection will send an error on write(), however, the way it's used with fmt.Printf(), I'm not sure if the error will propagate out.  We'll need to test it.

Reply:
@jsongHBO @jlutzhbo Doesn't this imply that the connection will be garbage collected and the fd subsequently closed?
```
// Fd returns the integer Unix file descriptor referencing the open file.
// ***The file descriptor is valid only until f.Close is called or f is garbage collected.***
// On Unix systems this will cause the SetDeadline methods to stop working.
func (f *File) Fd() uintptr {
	if f == nil {
		return ^(uintptr(0))
	}
```